In [1]:
import pandas as pd
from sklearn.model_selection import train_test_split

### CH05 표본 추출

In [2]:
df = pd.read_csv("bike.csv")
df.head(2)

datetime  season  holiday  workingday  weather  temp   atemp  \
0  2011-01-01 00:00:00       1        0           0        1  9.84  14.395   
1  2011-01-01 01:00:00       1        0           0        1  9.02  13.635   

   humidity  windspeed  casual  registered  count  
0        81        0.0       3          13     16  
1        80        0.0       8          32     40

In [7]:
df.sample(n = 2, random_state=123)

datetime  season  holiday  workingday  weather   temp  atemp  \
6495  2012-03-07 21:00:00       1        0           1        1  18.04  21.97   
7050  2012-04-12 03:00:00       2        0           1        1  13.12  15.15   

      humidity  windspeed  casual  registered  count  
6495        62    16.9979      11         177    188  
7050        61    16.9979       2           3      5

In [8]:
df["season"].unique()

array([1, 2, 3, 4], dtype=int64)

In [10]:
df.groupby("season").sample(n = 2, random_state=123)

datetime  season  holiday  workingday  weather   temp  \
336    2011-01-15 12:00:00       1        0           0        1   9.84   
184    2011-01-08 23:00:00       1        0           0        1   4.10   
7186   2012-04-17 19:00:00       2        0           1        1  24.60   
2250   2011-06-01 16:00:00       2        0           1        1  35.26   
3288   2011-08-06 22:00:00       3        0           0        2  28.70   
8959   2012-08-15 16:00:00       3        0           1        1  31.16   
4443   2011-10-17 04:00:00       4        0           1        1  21.32   
10584  2012-12-07 10:00:00       4        0           1        2  13.12   

        atemp  humidity  windspeed  casual  registered  count  
336    11.365        48    15.0013      15          74     89  
184     3.030        39    30.0026       0          22     22  
7186   30.305        26    16.9979      73         464    537  
2250   40.150        47    19.9995      39         227    266  
3288   33.335        84    22.0028      24          96    120  
8959   34.090        48    12.9980      91         340    431  
4443   25.000        59    11.0014       2           3      5  
10584  15.910        76    11.0014      16         150    166

In [11]:
df_train, df_test = train_test_split(df, train_size=0.8, random_state=123)
df_train.head(2)

datetime  season  holiday  workingday  weather   temp  \
9395  2012-09-14 20:00:00       3        0           1        1  27.06   
8023  2012-06-14 16:00:00       2        0           1        1  29.52   

       atemp  humidity  windspeed  casual  registered  count  
9395  31.060        61     7.0015      67         337    404  
8023  33.335        51    19.0012      70         368    438

In [13]:
df_test["temp"].max()

39.36

### CH06 데이터 전처리: 결측치, 이상치

In [14]:
df_6 = pd.read_csv("iris_missing.csv")
df_6.head(10)

Sepal_Length  Sepal_Width  Petal_Length  Petal_Width Species
0           5.1          NaN           NaN          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2     NaN
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2     NaN
5           5.4          3.9           1.7          0.4  setosa
6           NaN          NaN           NaN          NaN  setosa
7           5.0          3.4           NaN          0.2  setosa
8           4.4          2.9           1.4          0.2  setosa
9           4.9          3.1           NaN          0.1  setosa

In [15]:
df_6["Sepal_Length"].isna()

0      False
1      False
2      False
3      False
4      False
       ...  
145    False
146     True
147    False
148    False
149    False
Name: Sepal_Length, Length: 150, dtype: bool

In [16]:
df_6["Sepal_Length"].isna().sum()

7

In [17]:
df_6.isna().sum() # 각 변수별 결측치 개수

Sepal_Length     7
Sepal_Width     13
Petal_Length    14
Petal_Width      6
Species          2
dtype: int64

In [18]:
df_6.isna().sum(axis = 1) # 각 row 별 결측치 개수

0      2
1      0
2      1
3      0
4      1
      ..
145    0
146    3
147    0
148    0
149    1
Length: 150, dtype: int64

In [20]:
df_6.fillna(999)

Sepal_Length  Sepal_Width  Petal_Length  Petal_Width    Species
0             5.1        999.0         999.0          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2        999
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2        999
..            ...          ...           ...          ...        ...
145           6.7          3.0           5.2          2.3  virginica
146         999.0          2.5         999.0        999.0  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1        999.0  virginica

[150 rows x 5 columns]

In [21]:
df_6.fillna(value={"Sepal_Length":999})

Sepal_Length  Sepal_Width  Petal_Length  Petal_Width    Species
0             5.1          NaN           NaN          0.2     setosa
1             4.9          3.0           1.4          0.2     setosa
2             4.7          3.2           1.3          0.2        NaN
3             4.6          3.1           1.5          0.2     setosa
4             5.0          3.6           1.4          0.2        NaN
..            ...          ...           ...          ...        ...
145           6.7          3.0           5.2          2.3  virginica
146         999.0          2.5           NaN          NaN  virginica
147           6.5          3.0           5.2          2.0  virginica
148           6.2          3.4           5.4          2.3  virginica
149           5.9          3.0           5.1          NaN  virginica

[150 rows x 5 columns]

In [23]:
df_6_na = df_6.head(7)
df_6_na.head()

Sepal_Length  Sepal_Width  Petal_Length  Petal_Width Species
0           5.1          NaN           NaN          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2     NaN
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2     NaN

In [24]:
df_6_na = df_6_na.fillna(value={"Sepal_Width":df_6["Sepal_Width"].mean()})
df_6_na.head()

Sepal_Length  Sepal_Width  Petal_Length  Petal_Width Species
0           5.1     3.070073           NaN          0.2  setosa
1           4.9     3.000000           1.4          0.2  setosa
2           4.7     3.200000           1.3          0.2     NaN
3           4.6     3.100000           1.5          0.2  setosa
4           5.0     3.600000           1.4          0.2     NaN

In [25]:
df_6_na2 = df_6.head(10)
df_6_na2.head()

Sepal_Length  Sepal_Width  Petal_Length  Petal_Width Species
0           5.1          NaN           NaN          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2     NaN
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2     NaN

In [26]:
df_6_na2.dropna()

Sepal_Length  Sepal_Width  Petal_Length  Petal_Width Species
1           4.9          3.0           1.4          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
5           5.4          3.9           1.7          0.4  setosa
8           4.4          2.9           1.4          0.2  setosa

In [27]:
df_6["Sepal_Length"].median()

5.8

In [30]:
df_6["Sepal_Length"].quantile()

5.8

In [31]:
print(df_6["Sepal_Length"].quantile(0.25))
print(df_6["Sepal_Length"].quantile(0.75))

5.1
6.4


In [32]:
df_6.shape

(150, 5)

In [34]:
df_6_test=df_6["Sepal_Width"]
df_6_test = df_6_test.fillna(df_6_test.mean())
df_6_test.head()

0    3.070073
1    3.000000
2    3.200000
3    3.100000
4    3.600000
Name: Sepal_Width, dtype: float64

In [36]:
df_6_test.std() ** 2

0.17561945818840935

In [44]:
df_6 = pd.read_csv("iris_missing.csv")

In [46]:
sl_mean = df_6["Sepal_Length"].mean()
sl_std = df_6["Sepal_Length"].std()
print(sl_mean, sl_std)

5.8538461538461535 0.8235759566951737


In [49]:
cond_1 = df_6["Sepal_Length"] < (sl_mean - 1.5 * sl_std)
cond_2 = df_6["Sepal_Length"] > (sl_mean + 1.5 * sl_std)
df_out = df_6.loc[(cond_1) | (cond_2), ]
len(df_out)

19

In [55]:
df_6_test_3 = pd.read_csv("iris.csv")
df_6_test_3.head(2)

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width Species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa

In [56]:
t = df_6_test_3["Sepal.Length"].mean() + 1.5*df_6_test_3["Sepal.Length"].std()
k = df_6_test_3["Sepal.Length"].mean() - 1.5*df_6_test_3["Sepal.Length"].std()
print(t, k)

7.085432525300128 4.60123414136654


In [57]:
df_6_test_3.loc[(df_6_test_3["Sepal.Length"] > t), :].shape

(12, 5)

In [58]:
df_6_test_3.loc[(df_6_test_3["Sepal.Length"] < k), :].shape

(9, 5)

In [59]:
df_6_test_3.loc[(df_6_test_3["Sepal.Length"]>t)|(df_6_test_3["Sepal.Length"]<k), :].shape

(21, 5)

### CH07 데이터 전처리 : 파생변수 생성

In [61]:
import numpy as np

In [62]:
df_7 = pd.read_csv("iris.csv")
df_7.head(2)

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width Species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa

In [63]:
df_7["is_setosa"] = np.where(df_7["Species"] == "setosa", 1, 0)
df_7.head(2)

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width Species  is_setosa
0           5.1          3.5           1.4          0.2  setosa          1
1           4.9          3.0           1.4          0.2  setosa          1

In [65]:
pd.crosstab(df_7["Species"], df_7["is_setosa"])

is_setosa    0   1
Species           
setosa       0  50
versicolor  50   0
virginica   50   0

In [67]:
df_7["is_setosa2"] = (df_7["Species"] == "setosa" )+ 0
df_7.head()

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width Species  is_setosa  \
0           5.1          3.5           1.4          0.2  setosa          1   
1           4.9          3.0           1.4          0.2  setosa          1   
2           4.7          3.2           1.3          0.2  setosa          1   
3           4.6          3.1           1.5          0.2  setosa          1   
4           5.0          3.6           1.4          0.2  setosa          1   

   is_setosa2  
0           1  
1           1  
2           1  
3           1  
4           1

In [68]:
df_7 = df_7.rename(columns={"Sepal.Length":"SL"})
df_7.head(2)

SL  Sepal.Width  Petal.Length  Petal.Width Species  is_setosa  is_setosa2
0  5.1          3.5           1.4          0.2  setosa          1           1
1  4.9          3.0           1.4          0.2  setosa          1           1

In [69]:
bike = pd.read_csv("bike.csv")
bike.head(2)

datetime  season  holiday  workingday  weather  temp   atemp  \
0  2011-01-01 00:00:00       1        0           0        1  9.84  14.395   
1  2011-01-01 01:00:00       1        0           0        1  9.02  13.635   

   humidity  windspeed  casual  registered  count  
0        81        0.0       3          13     16  
1        80        0.0       8          32     40

In [72]:
bike_sub = bike.loc[:4, ["casual","registered"]]
bike_sub.head(2)

casual  registered
0       3          13
1       8          32

In [73]:
bike_sub.sum()

casual        19
registered    83
dtype: int64

In [74]:
bike_sub.sum(axis=1)

0    16
1    40
2    32
3    13
4     1
dtype: int64

In [75]:
bike_sub.apply(func=sum)

casual        19
registered    83
dtype: int64

In [76]:
bike_sub.apply(func=sum, axis = 1)

0    16
1    40
2    32
3    13
4     1
dtype: int64

In [77]:
bike_sub.apply(func = lambda x: round(x.mean()))

casual         4
registered    17
dtype: int64

In [78]:
bike_sub["casual"].astype("str")+ "대"

0    3대
1    8대
2    5대
3    3대
4    0대
Name: casual, dtype: object

In [79]:
bike["datetime"].str.slice(0,4)

0        2011
1        2011
2        2011
3        2011
4        2011
         ... 
10881    2012
10882    2012
10883    2012
10884    2012
10885    2012
Name: datetime, Length: 10886, dtype: object

In [80]:
bike_date = pd.to_datetime(bike["datetime"])
bike_date

0       2011-01-01 00:00:00
1       2011-01-01 01:00:00
2       2011-01-01 02:00:00
3       2011-01-01 03:00:00
4       2011-01-01 04:00:00
                ...        
10881   2012-12-19 19:00:00
10882   2012-12-19 20:00:00
10883   2012-12-19 21:00:00
10884   2012-12-19 22:00:00
10885   2012-12-19 23:00:00
Name: datetime, Length: 10886, dtype: datetime64[ns]

In [81]:
bike_date.dt.year

0        2011
1        2011
2        2011
3        2011
4        2011
         ... 
10881    2012
10882    2012
10883    2012
10884    2012
10885    2012
Name: datetime, Length: 10886, dtype: int64

In [82]:
bike_dum = pd.get_dummies(bike, columns=["season"])
bike_dum.head(2)

datetime  holiday  workingday  weather  temp   atemp  humidity  \
0  2011-01-01 00:00:00        0           0        1  9.84  14.395        81   
1  2011-01-01 01:00:00        0           0        1  9.02  13.635        80   

   windspeed  casual  registered  count  season_1  season_2  season_3  \
0        0.0       3          13     16         1         0         0   
1        0.0       8          32     40         1         0         0   

   season_4  
0         0  
1         0

In [84]:
bike_test = pd.read_csv("bike.csv")
bike_test.head(2)

datetime  season  holiday  workingday  weather  temp   atemp  \
0  2011-01-01 00:00:00       1        0           0        1  9.84  14.395   
1  2011-01-01 01:00:00       1        0           0        1  9.02  13.635   

   humidity  windspeed  casual  registered  count  
0        81        0.0       3          13     16  
1        80        0.0       8          32     40

In [85]:
bike_test = bike_test.loc[:, ["temp", "atemp"]]
bike_test.head(2)

temp   atemp
0  9.84  14.395
1  9.02  13.635

In [86]:
round(abs(bike_test["temp"] - bike_test["atemp"]).mean(), 3)

3.509

In [87]:
bike_test2 = pd.read_csv("bike.csv")
bike_test2.head(2)

datetime  season  holiday  workingday  weather  temp   atemp  \
0  2011-01-01 00:00:00       1        0           0        1  9.84  14.395   
1  2011-01-01 01:00:00       1        0           0        1  9.02  13.635   

   humidity  windspeed  casual  registered  count  
0        81        0.0       3          13     16  
1        80        0.0       8          32     40

In [89]:
bike_test2 = bike_test2.loc[bike_test2["casual"]>25, ]
bike_test2.head(2)

datetime  season  holiday  workingday  weather   temp   atemp  \
11  2011-01-01 11:00:00       1        0           0        1  14.76  16.665   
12  2011-01-01 12:00:00       1        0           0        1  17.22  21.210   

    humidity  windspeed  casual  registered  count  
11        81    19.0012      26          30     56  
12        77    19.0012      29          55     84

In [90]:
bike_test2["date"] = bike_test2["datetime"].str.slice(0, 10)
bike_test2.head(2)

datetime  season  holiday  workingday  weather   temp   atemp  \
11  2011-01-01 11:00:00       1        0           0        1  14.76  16.665   
12  2011-01-01 12:00:00       1        0           0        1  17.22  21.210   

    humidity  windspeed  casual  registered  count        date  
11        81    19.0012      26          30     56  2011-01-01  
12        77    19.0012      29          55     84  2011-01-01

In [91]:
bike_test2 = bike_test2.loc[:, ["date", "casual"]]
bike_test2.head(2)

date  casual
11  2011-01-01      26
12  2011-01-01      29

In [92]:
bike_test2.groupby("date")["casual"].sum()

date
2011-01-01    218
2011-01-15     88
2011-01-16     64
2011-02-06    264
2011-02-12    208
             ... 
2012-12-14    317
2012-12-15    622
2012-12-16    414
2012-12-18    321
2012-12-19    218
Name: casual, Length: 384, dtype: int64

In [94]:
bike_test3 = pd.read_csv("bike.csv")
bike_test3.head(2)

datetime  season  holiday  workingday  weather  temp   atemp  \
0  2011-01-01 00:00:00       1        0           0        1  9.84  14.395   
1  2011-01-01 01:00:00       1        0           0        1  9.02  13.635   

   humidity  windspeed  casual  registered  count  
0        81        0.0       3          13     16  
1        80        0.0       8          32     40

In [95]:
bike_test3["hour"] = pd.to_datetime(bike_test3["datetime"]).dt.hour
bike_test3.head(2)

datetime  season  holiday  workingday  weather  temp   atemp  \
0  2011-01-01 00:00:00       1        0           0        1  9.84  14.395   
1  2011-01-01 01:00:00       1        0           0        1  9.02  13.635   

   humidity  windspeed  casual  registered  count  hour  
0        81        0.0       3          13     16     0  
1        80        0.0       8          32     40     1

In [99]:
bike_test3_sub = bike_test3.groupby("hour")["registered"].mean().reset_index()
bike_test3_sub

hour  registered
0      0   44.826374
1      1   27.345815
2      2   18.080357
3      3    9.076212
4      4    5.144796
5      5   18.311947
6      6   72.109890
7      7  202.202198
8      8  341.226374
9      9  190.824176
10    10  128.973626
11    11  150.621978
12    12  187.677632
13    13  183.728070
14    14  166.853070
15    15  178.269737
16    16  241.289474
17    17  393.324561
18    18  369.462719
19    19  266.203947
20    20  191.785088
21    21  144.802632
22    22  110.973684
23    23   74.046053

In [100]:
bike_test3_sub["registered"].idxmax()

17

### CH08 데이터 전처리 : 데이터 병합

In [101]:
# pd.concat

In [102]:
bike = pd.read_csv("bike.csv")
bike.head(2)

datetime  season  holiday  workingday  weather  temp   atemp  \
0  2011-01-01 00:00:00       1        0           0        1  9.84  14.395   
1  2011-01-01 01:00:00       1        0           0        1  9.02  13.635   

   humidity  windspeed  casual  registered  count  
0        81        0.0       3          13     16  
1        80        0.0       8          32     40

In [106]:
bike1 = bike.iloc[:3, :4]
bike2 = bike.iloc[5:8, :4]

In [107]:
bike2

datetime  season  holiday  workingday
5  2011-01-01 05:00:00       1        0           0
6  2011-01-01 06:00:00       1        0           0
7  2011-01-01 07:00:00       1        0           0

In [108]:
bike3 = pd.concat([bike1, bike2])
bike3

datetime  season  holiday  workingday
0  2011-01-01 00:00:00       1        0           0
1  2011-01-01 01:00:00       1        0           0
2  2011-01-01 02:00:00       1        0           0
5  2011-01-01 05:00:00       1        0           0
6  2011-01-01 06:00:00       1        0           0
7  2011-01-01 07:00:00       1        0           0

In [109]:
# pd.merge

In [110]:
df_a = pd.read_csv("join_data_group_members.csv")
df_b = pd.read_csv("join_data_member_room.csv")

In [111]:
df_a.head(2)

team    No member
0  TWICE  1414     미나
1  TWICE  1462     나연

In [112]:
df_b.head(2)

floor  room name
0      3  3G-A   미나
1      3  3G-B   나연

In [113]:
pd.merge(left = df_a, right = df_b, left_on = "member", right_on = "name")

team    No member  floor  room name
0  TWICE  1414     미나      3  3G-A   미나
1  TWICE  1462     나연      3  3G-B   나연
2  TWICE  1178     다현      3  3G-C   다현
3  TWICE  1525     정연      4  4G-A   정연
4  TWICE  1194     지효      4  4G-B   지효
5  TWICE  1937     채연      4  4G-C   채연
6  TWICE  1817     모모      5  5G-A   모모
7  TWICE  1117     쯔위      5  5G-B   쯔위
8  TWICE  1298     사나      5  5G-C   사나

In [122]:
df_8_test2 = pd.read_csv("bike.csv")
df_8_test2.head(2)

datetime  season  holiday  workingday  weather  temp   atemp  \
0  2011-01-01 00:00:00       1        0           0        1  9.84  14.395   
1  2011-01-01 01:00:00       1        0           0        1  9.02  13.635   

   humidity  windspeed  casual  registered  count  
0        81        0.0       3          13     16  
1        80        0.0       8          32     40

In [123]:
df_8_test2["season"].value_counts()

4    2734
2    2733
3    2733
1    2686
Name: season, dtype: int64

In [124]:
df_8_test2 = df_8_test2.loc[df_8_test2["season"] % 2 == 0, ]
df_8_test2.head(2)

datetime  season  holiday  workingday  weather   temp  atemp  \
1323  2011-04-01 00:00:00       2        0           1        3  10.66  12.88   
1324  2011-04-01 01:00:00       2        0           1        3  10.66  12.88   

      humidity  windspeed  casual  registered  count  
1323       100    11.0014       0           6      6  
1324       100    11.0014       0           4      4

In [125]:
df_8_summer = df_8_test2.loc[df_8_test2["season"] == 2, ["datetime", "registered"]]
df_8_winter = df_8_test2.loc[df_8_test2["season"] == 4, ["datetime", "registered"]]

In [126]:
df_8_summer["hour"] = pd.to_datetime(df_8_summer["datetime"]).dt.hour
df_8_winter["hour"] = pd.to_datetime(df_8_winter["datetime"]).dt.hour

In [127]:
df_8_summer.head(2)

datetime  registered  hour
1323  2011-04-01 00:00:00           6     0
1324  2011-04-01 01:00:00           4     1

In [128]:
df_8_summer = df_8_summer.drop(columns="datetime")
df_8_summer.head(2)

registered  hour
1323           6     0
1324           4     1

In [129]:
df_8_winter = df_8_winter.drop(columns="datetime")
df_8_winter.head(2)

registered  hour
4055         106     0
4056          47     1

In [130]:
df_8_summer_sub = df_8_summer.groupby("hour")["registered"].mean().reset_index()
df_8_winter_sub = df_8_winter.groupby("hour")["registered"].mean().reset_index()

In [133]:
df_8_ans = pd.merge(left = df_8_summer_sub, right = df_8_winter_sub, left_on = "hour", right_on="hour")
df_8_ans.head(2)

hour  registered_x  registered_y
0     0     45.947368     50.017544
1     1     27.473684     30.842105

In [134]:
df_8_ans["ans"] = abs(df_8_ans["registered_x"] - df_8_ans["registered_y"])
df_8_ans["ans"].idxmax()

19

In [135]:
df_8_test3 = pd.read_csv("bike.csv")
df_8_test3.head(2)

datetime  season  holiday  workingday  weather  temp   atemp  \
0  2011-01-01 00:00:00       1        0           0        1  9.84  14.395   
1  2011-01-01 01:00:00       1        0           0        1  9.02  13.635   

   humidity  windspeed  casual  registered  count  
0        81        0.0       3          13     16  
1        80        0.0       8          32     40

In [137]:
df_8_test3["date"] = pd.to_datetime(df_8_test3["datetime"])
df_8_test3.head(2)

datetime  season  holiday  workingday  weather  temp   atemp  \
0  2011-01-01 00:00:00       1        0           0        1  9.84  14.395   
1  2011-01-01 01:00:00       1        0           0        1  9.02  13.635   

   humidity  windspeed  casual  registered  count                date  
0        81        0.0       3          13     16 2011-01-01 00:00:00  
1        80        0.0       8          32     40 2011-01-01 01:00:00

In [138]:
df_8_test3["dd"] = df_8_test3["date"].dt.date
df_8_test3.head(2)

datetime  season  holiday  workingday  weather  temp   atemp  \
0  2011-01-01 00:00:00       1        0           0        1  9.84  14.395   
1  2011-01-01 01:00:00       1        0           0        1  9.02  13.635   

   humidity  windspeed  casual  registered  count                date  \
0        81        0.0       3          13     16 2011-01-01 00:00:00   
1        80        0.0       8          32     40 2011-01-01 01:00:00   

           dd  
0  2011-01-01  
1  2011-01-01

In [139]:
df_8_test3["hour"] = df_8_test3["date"].dt.hour
df_8_test3.head(2)

datetime  season  holiday  workingday  weather  temp   atemp  \
0  2011-01-01 00:00:00       1        0           0        1  9.84  14.395   
1  2011-01-01 01:00:00       1        0           0        1  9.02  13.635   

   humidity  windspeed  casual  registered  count                date  \
0        81        0.0       3          13     16 2011-01-01 00:00:00   
1        80        0.0       8          32     40 2011-01-01 01:00:00   

           dd  hour  
0  2011-01-01     0  
1  2011-01-01     1

In [143]:
df_8_test3_rain = df_8_test3.groupby("dd")["humidity"].max().reset_index()
df_8_test3_rain.head(2)

dd  humidity
0  2011-01-01        94
1  2011-01-02       100

In [144]:
df_8_test3_rain = df_8_test3_rain.loc[df_8_test3_rain["humidity"] == 100, ]
df_8_test3_rain.head()


dd  humidity
1   2011-01-02       100
20  2011-02-02       100
23  2011-02-05       100
25  2011-02-07       100
42  2011-03-05       100

In [160]:
len(df_8_test3_rain)

40

In [153]:
df_8_test3_ans = pd.merge(left=df_8_test3_rain, right = df_8_test3.loc[:, ["dd", "temp", "count", "humidity", "hour"]], left_on= "dd", right_on="dd")

In [154]:
df_8_test3_ans.head(2)

dd  humidity_x   temp  count  humidity_y  hour
0  2011-01-02         100  18.86     17          88     0
1  2011-01-02         100  18.04     17          94     1

In [156]:
df_8_test3_ans = df_8_test3_ans.loc[df_8_test3_ans["temp"] > 30, ]
df_8_test3_ans.head(2)

dd  humidity_x   temp  count  humidity_y  hour
297  2011-04-11         100  30.34    141          48    15
298  2011-04-11         100  30.34    235          48    16

In [159]:
df_8_test3_ans.groupby("hour")["count"].mean().reset_index().mean()

hour      15.50
count    350.25
dtype: float64

In [161]:
df_8_test3_ans["count"].mean()

338.75

### CH09 데이터 전처리 : 정렬 및 변환

In [162]:
dia = pd.read_csv("diamonds.csv")
dia.head(2)

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31

In [163]:
pd.crosstab(dia["cut"], dia["clarity"])

clarity     I1    IF   SI1   SI2   VS1   VS2  VVS1  VVS2
cut                                                     
Fair       210     9   408   466   170   261    17    69
Good        96    71  1560  1081   648   978   186   286
Ideal      146  1212  4282  2598  3589  5071  2047  2606
Premium    205   230  3575  2949  1989  3357   616   870
Very Good   84   268  3240  2100  1775  2591   789  1235

In [164]:
pd.crosstab(dia["cut"], dia["clarity"], normalize=True)

clarity          I1        IF       SI1       SI2       VS1       VS2  \
cut                                                                     
Fair       0.003893  0.000167  0.007564  0.008639  0.003152  0.004839   
Good       0.001780  0.001316  0.028921  0.020041  0.012013  0.018131   
Ideal      0.002707  0.022469  0.079385  0.048165  0.066537  0.094012   
Premium    0.003801  0.004264  0.066277  0.054672  0.036874  0.062236   
Very Good  0.001557  0.004968  0.060067  0.038932  0.032907  0.048035   

clarity        VVS1      VVS2  
cut                            
Fair       0.000315  0.001279  
Good       0.003448  0.005302  
Ideal      0.037950  0.048313  
Premium    0.011420  0.016129  
Very Good  0.014627  0.022896

In [165]:
pd.crosstab(dia["cut"], dia["clarity"], normalize=True).round(2)

clarity     I1    IF   SI1   SI2   VS1   VS2  VVS1  VVS2
cut                                                     
Fair       0.0  0.00  0.01  0.01  0.00  0.00  0.00  0.00
Good       0.0  0.00  0.03  0.02  0.01  0.02  0.00  0.01
Ideal      0.0  0.02  0.08  0.05  0.07  0.09  0.04  0.05
Premium    0.0  0.00  0.07  0.05  0.04  0.06  0.01  0.02
Very Good  0.0  0.00  0.06  0.04  0.03  0.05  0.01  0.02

In [166]:
pd.crosstab(dia["cut"], dia["clarity"], values=dia["price"], aggfunc=pd.Series.mean)

clarity             I1           IF          SI1          SI2          VS1  \
cut                                                                          
Fair       3703.533333  1912.333333  4208.279412  5173.916309  4165.141176   
Good       3596.635417  4098.323944  3689.533333  4580.260870  3801.445988   
Ideal      4335.726027  2272.913366  3752.118169  4755.952656  3489.744497   
Premium    3947.331707  3856.143478  4455.269371  5545.936928  4485.462041   
Very Good  4078.226190  4396.216418  3932.391049  4988.688095  3805.353239   

clarity            VS2         VVS1         VVS2  
cut                                               
Fair       4174.724138  3871.352941  3349.768116  
Good       4262.236196  2254.774194  3079.108392  
Ideal      3284.550385  2468.129458  3250.290100  
Premium    4550.331248  2831.206169  3795.122989  
Very Good  4215.759552  2459.441065  3037.765182

In [168]:
dia.groupby(["cut", "clarity"])["price"].mean().reset_index()

cut clarity        price
0        Fair      I1  3703.533333
1        Fair      IF  1912.333333
2        Fair     SI1  4208.279412
3        Fair     SI2  5173.916309
4        Fair     VS1  4165.141176
5        Fair     VS2  4174.724138
6        Fair    VVS1  3871.352941
7        Fair    VVS2  3349.768116
8        Good      I1  3596.635417
9        Good      IF  4098.323944
10       Good     SI1  3689.533333
11       Good     SI2  4580.260870
12       Good     VS1  3801.445988
13       Good     VS2  4262.236196
14       Good    VVS1  2254.774194
15       Good    VVS2  3079.108392
16      Ideal      I1  4335.726027
17      Ideal      IF  2272.913366
18      Ideal     SI1  3752.118169
19      Ideal     SI2  4755.952656
20      Ideal     VS1  3489.744497
21      Ideal     VS2  3284.550385
22      Ideal    VVS1  2468.129458
23      Ideal    VVS2  3250.290100
24    Premium      I1  3947.331707
25    Premium      IF  3856.143478
26    Premium     SI1  4455.269371
27    Premium     SI2  5545.936928
28    Premium     VS1  4485.462041
29    Premium     VS2  4550.331248
30    Premium    VVS1  2831.206169
31    Premium    VVS2  3795.122989
32  Very Good      I1  4078.226190
33  Very Good      IF  4396.216418
34  Very Good     SI1  3932.391049
35  Very Good     SI2  4988.688095
36  Very Good     VS1  3805.353239
37  Very Good     VS2  4215.759552
38  Very Good    VVS1  2459.441065
39  Very Good    VVS2  3037.765182

In [169]:
dia_agg = dia.groupby(["cut", "clarity"])["price"].mean().reset_index()
dia_agg.head(2)

cut clarity        price
0  Fair      I1  3703.533333
1  Fair      IF  1912.333333

In [171]:
dia_agg.sort_values("price").head()

cut clarity        price
1        Fair      IF  1912.333333
14       Good    VVS1  2254.774194
17      Ideal      IF  2272.913366
38  Very Good    VVS1  2459.441065
22      Ideal    VVS1  2468.129458

In [172]:
dia_agg.sort_values("price", ascending=False).head()

cut clarity        price
27    Premium     SI2  5545.936928
3        Fair     SI2  5173.916309
35  Very Good     SI2  4988.688095
19      Ideal     SI2  4755.952656
11       Good     SI2  4580.260870

In [173]:
dia_agg.sort_values(["cut", "clarity"], ascending=[True, False]).head()

cut clarity        price
7  Fair    VVS2  3349.768116
6  Fair    VVS1  3871.352941
5  Fair     VS2  4174.724138
4  Fair     VS1  4165.141176
3  Fair     SI2  5173.916309

In [174]:
elec = pd.read_csv("elec_load_2017_7d.csv")
elec.head(2)

YEAR  MONTH  DAY   X1HR   X2HR   X3HR   X4HR   X5HR   X6HR   X7HR  ...  \
0  2017      1    1  60178  57862  56165  55135  54450  54401  54147  ...   
1  2017      1    2  57067  55148  53983  53571  54027  55388  57753  ...   

   X15HR  X16HR  X17HR  X18HR  X19HR  X20HR  X21HR  X22HR  X23HR  X24HR  
0  49213  49357  50362  53115  55205  55437  55348  55790  58195  59330  
1  72228  71945  72641  73387  72294  69989  67893  66504  67435  69275  

[2 rows x 27 columns]

In [175]:
elec_melt = elec.melt(id_vars=["YEAR", "MONTH", "DAY"])
elec_melt.head(2)

YEAR  MONTH  DAY variable  value
0  2017      1    1     X1HR  60178
1  2017      1    2     X1HR  57067

In [177]:
elec_pivot = elec_melt.pivot(index = ["YEAR", "MONTH", "DAY"], columns = "variable",values="value").reset_index()
elec_pivot.head()

variable  YEAR  MONTH  DAY  X10HR  X11HR  X12HR  X13HR  X14HR  X15HR  X16HR  \
0         2017      1    1  49572  50034  50026  49515  49365  49213  49357   
1         2017      1    2  70869  72867  73115  69165  71429  72228  71945   
2         2017      1    3  75999  76522  76019  72203  74725  75576  75168   
3         2017      1    4  76175  76375  75862  72288  74360  74977  74275   
4         2017      1    5  66806  68030  67878  64439  64986  64767  63613   

variable  ...  X23HR  X24HR   X2HR   X3HR   X4HR   X5HR   X6HR   X7HR   X8HR  \
0         ...  58195  59330  57862  56165  55135  54450  54401  54147  52467   
1         ...  67435  69275  55148  53983  53571  54027  55388  57753  60781   
2         ...  69593  71346  64233  63194  62461  62649  63396  65098  67745   
3         ...  68948  70834  66143  65051  64245  64152  64678  66494  68851   
4         ...  64350  66687  65801  64439  63437  63135  63250  63658  63441   

variable   X9HR  
0         50469  
1         66494  
2         72982  
3         73835  
4         65640  

[5 rows x 27 columns]

In [178]:
df_9_test1= pd.read_csv("bike.csv")
df_9_test1.head(2)

datetime  season  holiday  workingday  weather  temp   atemp  \
0  2011-01-01 00:00:00       1        0           0        1  9.84  14.395   
1  2011-01-01 01:00:00       1        0           0        1  9.02  13.635   

   humidity  windspeed  casual  registered  count  
0        81        0.0       3          13     16  
1        80        0.0       8          32     40

In [179]:
df_9_test1 = df_9_test1[["workingday", "holiday"]].copy()
df_9_test1.head(2)

workingday  holiday
0           0        0
1           0        0

In [180]:
pd.crosstab(df_9_test1["workingday"], df_9_test1["holiday"], normalize=True)

holiday            0         1
workingday                    
0           0.290557  0.028569
1           0.680875  0.000000

In [181]:
df_9_test2 = pd.read_csv("diamonds.csv")
df_9_test2.head(2)

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31

In [188]:
cross = pd.crosstab(df_9_test2["cut"], df_9_test2["color"])
cross

color         D     E     F     G     H     I    J
cut                                               
Fair        163   224   312   314   303   175  119
Good        662   933   909   871   702   522  307
Ideal      2834  3903  3826  4884  3115  2093  896
Premium    1603  2337  2331  2924  2360  1428  808
Very Good  1513  2400  2164  2299  1824  1204  678

In [189]:
cross.reset_index()

color        cut     D     E     F     G     H     I    J
0           Fair   163   224   312   314   303   175  119
1           Good   662   933   909   871   702   522  307
2          Ideal  2834  3903  3826  4884  3115  2093  896
3        Premium  1603  2337  2331  2924  2360  1428  808
4      Very Good  1513  2400  2164  2299  1824  1204  678

In [190]:
cross_melt = cross.reset_index().melt(id_vars="cut")
cross_melt

cut color  value
0        Fair     D    163
1        Good     D    662
2       Ideal     D   2834
3     Premium     D   1603
4   Very Good     D   1513
5        Fair     E    224
6        Good     E    933
7       Ideal     E   3903
8     Premium     E   2337
9   Very Good     E   2400
10       Fair     F    312
11       Good     F    909
12      Ideal     F   3826
13    Premium     F   2331
14  Very Good     F   2164
15       Fair     G    314
16       Good     G    871
17      Ideal     G   4884
18    Premium     G   2924
19  Very Good     G   2299
20       Fair     H    303
21       Good     H    702
22      Ideal     H   3115
23    Premium     H   2360
24  Very Good     H   1824
25       Fair     I    175
26       Good     I    522
27      Ideal     I   2093
28    Premium     I   1428
29  Very Good     I   1204
30       Fair     J    119
31       Good     J    307
32      Ideal     J    896
33    Premium     J    808
34  Very Good     J    678

In [193]:
cross_melt.sort_values("value", ascending=False)

cut color  value
17      Ideal     G   4884
7       Ideal     E   3903
12      Ideal     F   3826
22      Ideal     H   3115
18    Premium     G   2924
2       Ideal     D   2834
9   Very Good     E   2400
23    Premium     H   2360
8     Premium     E   2337
13    Premium     F   2331
19  Very Good     G   2299
14  Very Good     F   2164
27      Ideal     I   2093
24  Very Good     H   1824
3     Premium     D   1603
4   Very Good     D   1513
28    Premium     I   1428
29  Very Good     I   1204
6        Good     E    933
11       Good     F    909
32      Ideal     J    896
16       Good     G    871
33    Premium     J    808
21       Good     H    702
34  Very Good     J    678
1        Good     D    662
26       Good     I    522
15       Fair     G    314
10       Fair     F    312
31       Good     J    307
20       Fair     H    303
5        Fair     E    224
25       Fair     I    175
0        Fair     D    163
30       Fair     J    119

In [194]:
df_9_test3 = pd.read_csv("diamonds.csv")
df_9_test3.head(2)

carat      cut color clarity  depth  table  price     x     y     z
0   0.23    Ideal     E     SI2   61.5   55.0    326  3.95  3.98  2.43
1   0.21  Premium     E     SI1   59.8   61.0    326  3.89  3.84  2.31

In [200]:
df_9_test3_ans = df_9_test3.groupby(["cut", "color"])["price", "carat"].mean().reset_index()
df_9_test3_ans

C:\Users\GODIMODE\AppData\Local\Temp\ipykernel_12856\522755266.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  df_9_test3_ans = df_9_test3.groupby(["cut", "color"])["price", "carat"].mean().reset_index()


cut color        price     carat
0        Fair     D  4291.061350  0.920123
1        Fair     E  3682.312500  0.856607
2        Fair     F  3827.003205  0.904712
3        Fair     G  4239.254777  1.023822
4        Fair     H  5135.683168  1.219175
5        Fair     I  4685.445714  1.198057
6        Fair     J  4975.655462  1.341176
7        Good     D  3405.382175  0.744517
8        Good     E  3423.644159  0.745134
9        Good     F  3495.750275  0.775930
10       Good     G  4123.482204  0.850896
11       Good     H  4276.254986  0.914729
12       Good     I  5078.532567  1.057222
13       Good     J  4574.172638  1.099544
14      Ideal     D  2629.094566  0.565766
15      Ideal     E  2597.550090  0.578401
16      Ideal     F  3374.939362  0.655829
17      Ideal     G  3720.706388  0.700715
18      Ideal     H  3889.334831  0.799525
19      Ideal     I  4451.970377  0.913029
20      Ideal     J  4918.186384  1.063594
21    Premium     D  3631.292576  0.721547
22    Premium     E  3538.914420  0.717745
23    Premium     F  4324.890176  0.827036
24    Premium     G  4500.742134  0.841488
25    Premium     H  5216.706780  1.016449
26    Premium     I  5946.180672  1.144937
27    Premium     J  6294.591584  1.293094
28  Very Good     D  3470.467284  0.696424
29  Very Good     E  3214.652083  0.676317
30  Very Good     F  3778.820240  0.740961
31  Very Good     G  3872.753806  0.766799
32  Very Good     H  4535.390351  0.915948
33  Very Good     I  5255.879568  1.046952
34  Very Good     J  5103.513274  1.133215

In [201]:
df_9_test3_ans["ans"] = df_9_test3_ans["price"] / df_9_test3_ans["carat"]
df_9_test3_ans.head(2)

cut color       price     carat          ans
0  Fair     D  4291.06135  0.920123  4663.575143
1  Fair     E  3682.31250  0.856607  4298.717949

In [202]:
df_9_test3_ans.sort_values("ans", ascending=False)

cut color        price     carat          ans
24    Premium     G  4500.742134  0.841488  5348.553755
17      Ideal     G  3720.706388  0.700715  5309.874382
23    Premium     F  4324.890176  0.827036  5229.388117
26    Premium     I  5946.180672  1.144937  5193.456761
16      Ideal     F  3374.939362  0.655829  5146.069664
25    Premium     H  5216.706780  1.016449  5132.285040
30  Very Good     F  3778.820240  0.740961  5099.889612
31  Very Good     G  3872.753806  0.766799  5050.548821
21    Premium     D  3631.292576  0.721547  5032.648015
33  Very Good     I  5255.879568  1.046952  5020.173260
28  Very Good     D  3470.467284  0.696424  4983.265477
32  Very Good     H  4535.390351  0.915948  4951.578090
22    Premium     E  3538.914420  0.717745  4930.601477
19      Ideal     I  4451.970377  0.913029  4876.044103
27    Premium     J  6294.591584  1.293094  4867.852836
18      Ideal     H  3889.334831  0.799525  4864.557602
10       Good     G  4123.482204  0.850896  4846.049951
12       Good     I  5078.532567  1.057222  4803.656658
29  Very Good     E  3214.652083  0.676317  4753.175904
11       Good     H  4276.254986  0.914729  4674.885539
0        Fair     D  4291.061350  0.920123  4663.575143
14      Ideal     D  2629.094566  0.565766  4646.967032
20      Ideal     J  4918.186384  1.063594  4624.121178
8        Good     E  3423.644159  0.745134  4594.669237
7        Good     D  3405.382175  0.744517  4573.950535
9        Good     F  3495.750275  0.775930  4505.241592
34  Very Good     J  5103.513274  1.133215  4503.568825
15      Ideal     E  2597.550090  0.578401  4490.913843
1        Fair     E  3682.312500  0.856607  4298.717949
2        Fair     F  3827.003205  0.904712  4230.081128
4        Fair     H  5135.683168  1.219175  4212.425219
13       Good     J  4574.172638  1.099544  4160.063396
3        Fair     G  4239.254777  1.023822  4140.618390
5        Fair     I  4685.445714  1.198057  3910.869980
6        Fair     J  4975.655462  1.341176  3709.918546

### CH11 상관 분석

In [2]:
import pandas as pd
from scipy.stats import pearsonr
from scipy.stats import spearmanr
from scipy.stats import kendalltau

In [3]:
df = pd.read_csv("bike.csv")
df.head(2)

datetime  season  holiday  workingday  weather  temp   atemp  \
0  2011-01-01 00:00:00       1        0           0        1  9.84  14.395   
1  2011-01-01 01:00:00       1        0           0        1  9.02  13.635   

   humidity  windspeed  casual  registered  count  
0        81        0.0       3          13     16  
1        80        0.0       8          32     40

In [4]:
df.corr()

C:\Users\GODIMODE\AppData\Local\Temp\ipykernel_4168\1134722465.py:1: FutureWarning: The default value of numeric_only in DataFrame.corr is deprecated. In a future version, it will default to False. Select only valid columns or specify the value of numeric_only to silence this warning.
  df.corr()


season   holiday  workingday   weather      temp     atemp  \
season      1.000000  0.029368   -0.008126  0.008879  0.258689  0.264744   
holiday     0.029368  1.000000   -0.250491 -0.007074  0.000295 -0.005215   
workingday -0.008126 -0.250491    1.000000  0.033772  0.029966  0.024660   
weather     0.008879 -0.007074    0.033772  1.000000 -0.055035 -0.055376   
temp        0.258689  0.000295    0.029966 -0.055035  1.000000  0.984948   
atemp       0.264744 -0.005215    0.024660 -0.055376  0.984948  1.000000   
humidity    0.190610  0.001929   -0.010880  0.406244 -0.064949 -0.043536   
windspeed  -0.147121  0.008409    0.013373  0.007261 -0.017852 -0.057473   
casual      0.096758  0.043799   -0.319111 -0.135918  0.467097  0.462067   
registered  0.164011 -0.020956    0.119460 -0.109340  0.318571  0.314635   
count       0.163439 -0.005393    0.011594 -0.128655  0.394454  0.389784   

            humidity  windspeed    casual  registered     count  
season      0.190610  -0.147121  0.096758    0.164011  0.163439  
holiday     0.001929   0.008409  0.043799   -0.020956 -0.005393  
workingday -0.010880   0.013373 -0.319111    0.119460  0.011594  
weather     0.406244   0.007261 -0.135918   -0.109340 -0.128655  
temp       -0.064949  -0.017852  0.467097    0.318571  0.394454  
atemp      -0.043536  -0.057473  0.462067    0.314635  0.389784  
humidity    1.000000  -0.318607 -0.348187   -0.265458 -0.317371  
windspeed  -0.318607   1.000000  0.092276    0.091052  0.101369  
casual     -0.348187   0.092276  1.000000    0.497250  0.690414  
registered -0.265458   0.091052  0.497250    1.000000  0.970948  
count      -0.317371   0.101369  0.690414    0.970948  1.000000

In [5]:
df[["casual", "registered", "count"]].corr(method = "spearman")

casual  registered     count
casual      1.000000    0.775785  0.847378
registered  0.775785    1.000000  0.988901
count       0.847378    0.988901  1.000000

In [6]:
df_11_test1 = pd.read_csv("bike.csv")
df_11_test1.head(2)

datetime  season  holiday  workingday  weather  temp   atemp  \
0  2011-01-01 00:00:00       1        0           0        1  9.84  14.395   
1  2011-01-01 01:00:00       1        0           0        1  9.02  13.635   

   humidity  windspeed  casual  registered  count  
0        81        0.0       3          13     16  
1        80        0.0       8          32     40

In [8]:
df_11_test1[["temp", "atemp", "humidity", "casual"]].corr()

temp     atemp  humidity    casual
temp      1.000000  0.984948 -0.064949  0.467097
atemp     0.984948  1.000000 -0.043536  0.462067
humidity -0.064949 -0.043536  1.000000 -0.348187
casual    0.467097  0.462067 -0.348187  1.000000

In [19]:
df_11_test2 = df_11_test1[["season", "atemp", "casual"]].copy()
df_11_test2.head(2)

season   atemp  casual
0       1  14.395       3
1       1  13.635       8

In [20]:
df_11_test2.loc[df_11_test2["season"] == 1, ].corr()

season     atemp    casual
season     NaN       NaN       NaN
atemp      NaN  1.000000  0.478312
casual     NaN  0.478312  1.000000

In [21]:
df_11_test2.loc[df_11_test2["season"] == 2, ].corr()

season     atemp    casual
season     NaN       NaN       NaN
atemp      NaN  1.000000  0.378122
casual     NaN  0.378122  1.000000

In [22]:
df_11_test2.loc[df_11_test2["season"] == 3, ].corr()

season     atemp    casual
season     NaN       NaN       NaN
atemp      NaN  1.000000  0.381423
casual     NaN  0.381423  1.000000

In [23]:
df_11_test2.loc[df_11_test2["season"] == 4, ].corr()

season     atemp    casual
season     NaN       NaN       NaN
atemp      NaN  1.000000  0.443751
casual     NaN  0.443751  1.000000

In [24]:
df_11_test2.groupby("season")[["atemp", "casual"]].corr()

atemp    casual
season                           
1      atemp   1.000000  0.478312
       casual  0.478312  1.000000
2      atemp   1.000000  0.378122
       casual  0.378122  1.000000
3      atemp   1.000000  0.381423
       casual  0.381423  1.000000
4      atemp   1.000000  0.443751
       casual  0.443751  1.000000

In [29]:
df_11_test3 = df_11_test1[["weather", "temp", "casual"]].copy()
df_11_test3.head(2)

weather  temp  casual
0        1  9.84       3
1        1  9.02       8

In [30]:
df_11_test3["weather"].unique()

array([1, 2, 3, 4], dtype=int64)

In [31]:
df_11_test3.loc[df_11_test3["weather"]==1, ].corr()

weather      temp    casual
weather      NaN       NaN       NaN
temp         NaN  1.000000  0.471053
casual       NaN  0.471053  1.000000

In [32]:
df_11_test3.loc[df_11_test3["weather"]!=1, ].corr()

weather      temp    casual
weather  1.000000 -0.005939 -0.137405
temp    -0.005939  1.000000  0.446361
casual  -0.137405  0.446361  1.000000

In [34]:
round(abs(0.471053-0.446361), 3)

0.025

In [35]:
df_11_test3["is_sunny"] = (df_11_test3["weather"] == 1) + 0
df_11_test3.head(2)

weather  temp  casual  is_sunny
0        1  9.84       3         1
1        1  9.02       8         1

In [36]:
df_11_test3.groupby("is_sunny")[["temp", "casual"]].corr()

temp    casual
is_sunny                           
0        temp    1.000000  0.446361
         casual  0.446361  1.000000
1        temp    1.000000  0.471053
         casual  0.471053  1.000000

### CH12 비계층적 군집분석

In [37]:
from sklearn.cluster import KMeans
from sklearn.preprocessing import MinMaxScaler
from sklearn.preprocessing import StandardScaler

In [38]:
df12 = pd.read_csv("iris.csv")
df12.head(2)

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width Species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa

In [40]:
df12_1 = df12.head()
df12_2 = df12.tail(1)

In [42]:
nor_model = MinMaxScaler().fit(df12_1.iloc[:, :-1])
nor_model

MinMaxScaler()

In [44]:
df12_1_nor = nor_model.transform(df12_1.iloc[:, :-1])
df12_1_nor

array([[1.        , 0.83333333, 0.5       , 0.        ],
       [0.6       , 0.        , 0.5       , 0.        ],
       [0.2       , 0.33333333, 0.        , 0.        ],
       [0.        , 0.16666667, 1.        , 0.        ],
       [0.8       , 1.        , 0.5       , 0.        ]])

In [46]:
df12_1_nor = pd.DataFrame(df12_1_nor, columns=df12_1.columns[:-1])
df12_1_nor

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width
0           1.0     0.833333           0.5          0.0
1           0.6     0.000000           0.5          0.0
2           0.2     0.333333           0.0          0.0
3           0.0     0.166667           1.0          0.0
4           0.8     1.000000           0.5          0.0

In [47]:
model_kmeans = KMeans(n_clusters=3, random_state=123).fit(df12.iloc[:, :-1])
model_kmeans

C:\Users\GODIMODE\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


KMeans(n_clusters=3, random_state=123)

In [50]:
model_kmeans.labels_

array([1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1,
       1, 1, 1, 1, 1, 1, 2, 2, 0, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 0, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2,
       2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 2, 0, 2, 0, 0, 0, 0, 2, 0, 0, 0,
       0, 0, 0, 2, 2, 0, 0, 0, 0, 2, 0, 2, 0, 2, 0, 0, 2, 2, 0, 0, 0, 0,
       0, 2, 0, 0, 0, 0, 2, 0, 0, 0, 2, 0, 0, 0, 2, 0, 0, 2])

In [51]:
model_kmeans.cluster_centers_

array([[6.85      , 3.07368421, 5.74210526, 2.07105263],
       [5.006     , 3.428     , 1.462     , 0.246     ],
       [5.9016129 , 2.7483871 , 4.39354839, 1.43387097]])

In [53]:
df12["cluster"] = model_kmeans.labels_
df12.groupby("cluster").mean()

C:\Users\GODIMODE\AppData\Local\Temp\ipykernel_4168\1345592610.py:2: FutureWarning: The default value of numeric_only in DataFrameGroupBy.mean is deprecated. In a future version, numeric_only will default to False. Either specify numeric_only or select only columns which should be valid for the function.
  df12.groupby("cluster").mean()


Sepal.Length  Sepal.Width  Petal.Length  Petal.Width
cluster                                                      
0            6.850000     3.073684      5.742105     2.071053
1            5.006000     3.428000      1.462000     0.246000
2            5.901613     2.748387      4.393548     1.433871

In [54]:
df12_test1 = pd.read_csv("diabetes.csv")
df12_test1.head(2)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0

In [55]:
df12_test1 = df12_test1.loc[df12_test1["BMI"]!=0, ]
df12_test1.head(2)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0

In [56]:
model = KMeans(n_clusters=4, random_state=123).fit(df12_test1)
model

C:\Users\GODIMODE\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


KMeans(n_clusters=4, random_state=123)

In [58]:
df12_test1["cluster"] = model.labels_
df12_test1.head(2)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   

   DiabetesPedigreeFunction  Age  Outcome  cluster  
0                     0.627   50        1        0  
1                     0.351   31        0        0

In [59]:
df12_test1["cluster"].value_counts()

0    407
2    212
3    114
1     24
Name: cluster, dtype: int64

In [60]:
df12_test1.groupby("cluster")["Insulin"].mean()

cluster
0      4.103194
1    509.166667
2    102.674528
3    224.035088
Name: Insulin, dtype: float64

In [62]:
df12_test2 = pd.read_csv("diabetes.csv")
df12_test2.head(2)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0

In [63]:
df12_test2 = df12_test2.loc[df12_test2["BMI"] != 0, ]

In [64]:
nor_model = MinMaxScaler().fit(df12_test2)
nor_model

MinMaxScaler()

In [65]:
nor_df12_test2 = nor_model.transform(df12_test2)
nor_df12_test2

array([[0.35294118, 0.74371859, 0.59016393, ..., 0.23441503, 0.48333333,
        1.        ],
       [0.05882353, 0.42713568, 0.54098361, ..., 0.11656704, 0.16666667,
        0.        ],
       [0.47058824, 0.91959799, 0.52459016, ..., 0.25362938, 0.18333333,
        1.        ],
       ...,
       [0.29411765, 0.6080402 , 0.59016393, ..., 0.07130658, 0.15      ,
        0.        ],
       [0.05882353, 0.63316583, 0.49180328, ..., 0.11571307, 0.43333333,
        1.        ],
       [0.05882353, 0.46733668, 0.57377049, ..., 0.10119556, 0.03333333,
        0.        ]])

In [67]:
nor_df12_test2 = pd.DataFrame(nor_df12_test2, columns=df12_test2.columns)
nor_df12_test2.head(2)

Pregnancies   Glucose  BloodPressure  SkinThickness  Insulin       BMI  \
0     0.352941  0.743719       0.590164       0.353535      0.0  0.314928   
1     0.058824  0.427136       0.540984       0.292929      0.0  0.171779   

   DiabetesPedigreeFunction       Age  Outcome  
0                  0.234415  0.483333      1.0  
1                  0.116567  0.166667      0.0

In [68]:
kModel = KMeans(n_clusters=4, random_state=123).fit(nor_df12_test2)
df12_test2["cluster"] = kModel.labels_
df12_test2.head(2)

C:\Users\GODIMODE\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   

   DiabetesPedigreeFunction  Age  Outcome  cluster  
0                     0.627   50        1        2  
1                     0.351   31        0        0

In [78]:
df12_test2["cluster"].value_counts()

0    361
1    135
2    131
3    130
Name: cluster, dtype: int64

In [79]:
df12_test2.groupby("cluster")["Age"].mean()

cluster
0    25.667590
1    29.977778
2    44.297710
3    46.753846
Name: Age, dtype: float64

In [89]:
df12_test3 = pd.read_csv("diabetes.csv")
df12_test3 = df12_test3.loc[df12_test3["BMI"] != 0, ]
df12_test3.head(2)

Pregnancies  Glucose  BloodPressure  SkinThickness  Insulin   BMI  \
0            6      148             72             35        0  33.6   
1            1       85             66             29        0  26.6   

   DiabetesPedigreeFunction  Age  Outcome  
0                     0.627   50        1  
1                     0.351   31        0

In [91]:
kModel3 = KMeans(n_clusters=3, random_state=123).fit(df12_test3)

C:\Users\GODIMODE\AppData\Local\Programs\Python\Python39\lib\site-packages\sklearn\cluster\_kmeans.py:870: FutureWarning: The default value of `n_init` will change from 10 to 'auto' in 1.4. Set the value of `n_init` explicitly to suppress the warning
  warnings.warn(


In [92]:
kModel3

KMeans(n_clusters=3, random_state=123)

In [94]:
df_center = pd.DataFrame(kModel3.cluster_centers_, columns=df12_test3.columns)
df_center

Pregnancies     Glucose  BloodPressure  SkinThickness     Insulin  \
0     4.026316  158.447368      72.000000      32.263158  441.289474   
1     3.975258  114.237113      68.647423      15.259794   14.696907   
2     3.542735  129.376068      71.478632      30.337607  159.401709   

         BMI  DiabetesPedigreeFunction        Age   Outcome  
0  35.107895                  0.569211  34.763158  0.578947  
1  31.440619                  0.434579  33.808247  0.301031  
2  34.134615                  0.535188  31.948718  0.418803

In [95]:
df_center = df_center.transpose()
df_center

0           1           2
Pregnancies                 4.026316    3.975258    3.542735
Glucose                   158.447368  114.237113  129.376068
BloodPressure              72.000000   68.647423   71.478632
SkinThickness              32.263158   15.259794   30.337607
Insulin                   441.289474   14.696907  159.401709
BMI                        35.107895   31.440619   34.134615
DiabetesPedigreeFunction    0.569211    0.434579    0.535188
Age                        34.763158   33.808247   31.948718
Outcome                     0.578947    0.301031    0.418803

In [99]:
(((df_center[0] - df_center[1]) ** 2).sum()) ** 0.5

429.24419310888464

In [100]:
(((df_center[1] - df_center[2]) ** 2).sum()) ** 0.5

146.33847909815492

In [101]:
(((df_center[0] - df_center[2]) ** 2).sum()) ** 0.5

283.405999774738

### CH13 단순 회귀 분석

In [102]:
from statsmodels.formula.api import ols

In [103]:
df13 = pd.read_csv("iris.csv")
df13.head(2)

Sepal.Length  Sepal.Width  Petal.Length  Petal.Width Species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa

In [104]:
df13.columns = ["SL", "SW", "PL", "PW", "species"]
df13.head(2)

SL   SW   PL   PW species
0  5.1  3.5  1.4  0.2  setosa
1  4.9  3.0  1.4  0.2  setosa

In [105]:
model = ols(formula = "SL ~ SW", data=df13).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     SL   R-squared:                       0.014
Model:                            OLS   Adj. R-squared:                  0.007
Method:                 Least Squares   F-statistic:                     2.074
Date:                Sun, 02 Apr 2023   Prob (F-statistic):              0.152
Time:                        21:26:00   Log-Likelihood:                -183.00
No. Observations:                 150   AIC:                             370.0
Df Residuals:                     148   BIC:                             376.0
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      6.5262      0.479     13.628      0.000       5.580       7.473
SW            -0.2234      0.155     -1.440      0.152      -0.530       0.083
==============================================================================
Omnibus:                        4.389   Durbin-Watson:                   0.952
Prob(Omnibus):                  0.111   Jarque-Bera (JB):                4.237
Skew:                           0.360   Prob(JB):                        0.120
Kurtosis:                       2.600   Cond. No.                         24.2
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [106]:
model = ols(formula = "PL ~ PW", data=df13).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                     PL   R-squared:                       0.927
Model:                            OLS   Adj. R-squared:                  0.927
Method:                 Least Squares   F-statistic:                     1882.
Date:                Sun, 02 Apr 2023   Prob (F-statistic):           4.68e-86
Time:                        21:27:53   Log-Likelihood:                -101.18
No. Observations:                 150   AIC:                             206.4
Df Residuals:                     148   BIC:                             212.4
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept      1.0836      0.073     14.850      0.000       0.939       1.228
PW             2.2299      0.051     43.387      0.000       2.128       2.332
==============================================================================
Omnibus:                        2.438   Durbin-Watson:                   1.430
Prob(Omnibus):                  0.295   Jarque-Bera (JB):                1.966
Skew:                           0.211   Prob(JB):                        0.374
Kurtosis:                       3.369   Cond. No.                         3.70
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

R 계수 -> 1에 가까울수록 좋다??

P-value -> 0 에 가까울수록 좋다

coef에서 PW 는 기울기, Intercept 는 절편

y = 2.2299x + 1.0836

In [108]:
df13["pred"] = model.predict(df13)
df13.head()

SL   SW   PL   PW species      pred
0  5.1  3.5  1.4  0.2  setosa  1.529546
1  4.9  3.0  1.4  0.2  setosa  1.529546
2  4.7  3.2  1.3  0.2  setosa  1.529546
3  4.6  3.1  1.5  0.2  setosa  1.529546
4  5.0  3.6  1.4  0.2  setosa  1.529546

In [110]:
from sklearn.linear_model import LinearRegression

In [111]:
model = LinearRegression().fit(X = df13[["PL"]], 
                              y = df13["PW"])
model

LinearRegression()

In [112]:
model.coef_

array([0.41575542])

In [113]:
model.intercept_

-0.3630755213190291

In [115]:
model.predict(df13[["PL"]])

array([0.21898206, 0.21898206, 0.17740652, 0.2605576 , 0.21898206,
       0.34370869, 0.21898206, 0.2605576 , 0.21898206, 0.2605576 ,
       0.2605576 , 0.30213314, 0.21898206, 0.09425544, 0.13583098,
       0.2605576 , 0.17740652, 0.21898206, 0.34370869, 0.2605576 ,
       0.34370869, 0.2605576 , 0.0526799 , 0.34370869, 0.42685977,
       0.30213314, 0.30213314, 0.2605576 , 0.21898206, 0.30213314,
       0.30213314, 0.2605576 , 0.2605576 , 0.21898206, 0.2605576 ,
       0.13583098, 0.17740652, 0.21898206, 0.17740652, 0.2605576 ,
       0.17740652, 0.17740652, 0.17740652, 0.30213314, 0.42685977,
       0.21898206, 0.30213314, 0.21898206, 0.2605576 , 0.21898206,
       1.59097494, 1.50782385, 1.67412602, 1.29994614, 1.54939939,
       1.50782385, 1.59097494, 1.00891735, 1.54939939, 1.2583706 ,
       1.09206844, 1.38309723, 1.29994614, 1.59097494, 1.13364398,
       1.46624831, 1.50782385, 1.34152169, 1.50782385, 1.2583706 ,
       1.63255048, 1.29994614, 1.67412602, 1.59097494, 1.42467

In [116]:
from sklearn.metrics import mean_absolute_error
from sklearn.metrics import mean_squared_error

In [117]:
mean_absolute_error(y_true = df13["PL"], y_pred = df13["PW"])

2.558666666666667

In [118]:
mean_squared_error(y_true=df13["PL"], y_pred=df13["PW"])

7.645466666666667

In [119]:
mean_squared_error(y_true=df13["PL"], y_pred=df13["PW"]) ** 0.5

2.76504370067937

In [120]:
from statsmodels.formula.api import ols

In [155]:
df13_test1 = pd.read_csv("bike.csv")
df13_test1.head(2)

datetime  season  holiday  workingday  weather  temp   atemp  \
0  2011-01-01 00:00:00       1        0           0        1  9.84  14.395   
1  2011-01-01 01:00:00       1        0           0        1  9.02  13.635   

   humidity  windspeed  casual  registered  count  
0        81        0.0       3          13     16  
1        80        0.0       8          32     40

In [122]:
from sklearn.model_selection import train_test_split

In [156]:
train, test = train_test_split(df13_test1, train_size=0.7, random_state=123)

In [127]:
model = ols(formula="registered ~ temp", data = train).fit()

In [128]:
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:             registered   R-squared:                       0.106
Model:                            OLS   Adj. R-squared:                  0.106
Method:                 Least Squares   F-statistic:                     902.3
Date:                Sun, 02 Apr 2023   Prob (F-statistic):          1.92e-187
Time:                        21:48:06   Log-Likelihood:                -48650.
No. Observations:                7620   AIC:                         9.730e+04
Df Residuals:                    7618   BIC:                         9.732e+04
Df Model:                           1                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
Intercept     27.5151      4.559      6.036      0.000      18.579      36.452
temp           6.3391      0.211     30.038      0.000       5.925       6.753
==============================================================================
Omnibus:                     2097.525   Durbin-Watson:                   2.022
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             5337.402
Skew:                           1.502   Prob(JB):                         0.00
Kurtosis:                       5.790   Cond. No.                         60.1
==============================================================================

Notes:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

In [129]:
model = ols(formula="casual ~ atemp", data = train).fit()
model

In [136]:
test

datetime  season  holiday  workingday  weather   temp  \
6495  2012-03-07 21:00:00       1        0           1        1  18.04   
7050  2012-04-12 03:00:00       2        0           1        1  13.12   
558   2011-02-06 10:00:00       1        0           0        1  12.30   
5085  2011-12-05 23:00:00       4        0           1        2  18.86   
3328  2011-08-08 14:00:00       3        0           1        1  33.62   
...                   ...     ...      ...         ...      ...    ...   
3733  2011-09-06 12:00:00       3        0           1        3  22.14   
24    2011-01-02 00:00:00       1        0           0        2  18.86   
9650  2012-10-06 11:00:00       4        0           0        1  26.24   
6246  2012-02-16 12:00:00       1        0           1        3  13.12   
5239  2011-12-12 09:00:00       4        0           1        2   7.38   

       atemp  humidity  windspeed  casual  registered  count  
6495  21.970        62    16.9979      11         177    188  
7050  15.150        61    16.9979       2           3      5  
558   14.395        52    16.9979      15          74     89  
5085  22.725        88     8.9981       8          54     62  
3328  37.120        43    19.9995      41         109    150  
...      ...       ...        ...     ...         ...    ...  
3733  25.760        83    26.0027      16          51     67  
24    22.725        88    19.9995       4          13     17  
9650  30.305        65    16.9979     258         389    647  
6246  16.665        76     6.0032       1          44     45  
5239  10.605        74     6.0032      12         208    220  

[3266 rows x 12 columns]

In [137]:
pred = model.predict(test)
pred

6495    31.499001
7050    12.626390
558     10.537120
5085    33.588271
3328    73.422763
          ...    
3733    41.986859
24      33.588271
9650    54.563988
6246    16.818766
5239     0.049261
Length: 3266, dtype: float64

In [132]:
from sklearn.metrics import mean_squared_error

In [138]:
mean_squared_error(y_true = test["casual"], 
                  y_pred=pred) **0.5

44.46237010271433

In [139]:
from statsmodels.formula.api import ols

In [157]:
model = ols(formula="casual ~ atemp", data = train).fit()
model

In [158]:
pred = model.predict(test)
pred

6495    31.499001
7050    12.626390
558     10.537120
5085    33.588271
3328    73.422763
          ...    
3733    41.986859
24      33.588271
9650    54.563988
6246    16.818766
5239     0.049261
Length: 3266, dtype: float64

In [159]:
test["pred"] = pred
test.head(2)

datetime  season  holiday  workingday  weather   temp  atemp  \
6495  2012-03-07 21:00:00       1        0           1        1  18.04  21.97   
7050  2012-04-12 03:00:00       2        0           1        1  13.12  15.15   

      humidity  windspeed  casual  registered  count       pred  
6495        62    16.9979      11         177    188  31.499001  
7050        61    16.9979       2           3      5  12.626390

In [145]:
from sklearn.metrics import mean_squared_error

In [160]:
summer_y_true = test.loc[test["season"] == 2, ["casual", "pred"]]
winter_y_true = test.loc[test["season"] == 4, ["casual", "pred"]]

In [161]:
mean_squared_error(y_true=summer_y_true["casual"], y_pred=summer_y_true["pred"]) ** 0.5

55.16232790957267

In [162]:
mean_squared_error(y_true=winter_y_true["casual"], y_pred=winter_y_true["pred"]) ** 0.5

40.57002529173707

In [154]:
round(abs(55.16232790957267 - 40.57002529173707), 2)

14.59

In [163]:
df = pd.read_csv("bike.csv")

In [164]:
df_s2= df.loc[df["season"] == 2,]
df_s4= df.loc[df["season"] == 4,]

In [165]:
df_s2_train, df_s2_test = train_test_split(df_s2, train_size=0.7, random_state=123)
df_s4_train, df_s4_test = train_test_split(df_s4, train_size=0.7, random_state=123)

In [166]:
model_s2 = ols(formula="casual ~ atemp", data = df_s2_train).fit()
model_s4 = ols(formula="casual ~ atemp", data = df_s4_train).fit()

In [167]:
pred_s2 = model_s2.predict(df_s2_test)
pred_s4 = model_s4.predict(df_s4_test)
a2 = mean_squared_error(y_pred = pred_s2, y_true = df_s2_test["casual"]) ** 0.5
a4 = mean_squared_error(y_pred = pred_s4, y_true = df_s4_test["casual"]) ** 0.5

In [177]:
abs(a2 - a4)

8.648423450414171

In [169]:
df_train, df_test = train_test_split(df, train_size=0.7, random_state=123)

In [170]:
model = ols(formula="casual ~ atemp", data = df_train).fit()

In [171]:
pred = model.predict(df_test)

In [172]:
df_test["pred"] = pred
df_test.head(2)

datetime  season  holiday  workingday  weather   temp  atemp  \
6495  2012-03-07 21:00:00       1        0           1        1  18.04  21.97   
7050  2012-04-12 03:00:00       2        0           1        1  13.12  15.15   

      humidity  windspeed  casual  registered  count       pred  
6495        62    16.9979      11         177    188  31.499001  
7050        61    16.9979       2           3      5  12.626390

In [173]:
df_test_summer =df_test.loc[df_test["season"] == 2, ]
df_test_winter =df_test.loc[df_test["season"] == 4, ]

In [175]:
ans2 = mean_squared_error(y_true=df_test_summer["casual"], y_pred= df_test_summer["pred"]) ** 0.5
ans4 = mean_squared_error(y_true=df_test_winter["casual"], y_pred= df_test_winter["pred"]) ** 0.5

In [176]:
round(abs(ans2 - ans4), 2)

14.59